std/math.d

```diff
@@ -1849,10 +1883,28 @@ real tanh(real x) @safe pure nothrow @nogc
 }
 
 /// ditto
-double tanh(double x) @safe pure nothrow @nogc { return tanh(cast(real) x); }
+double tanh(double x) @safe pure nothrow @nogc
+{
+    if (fabs(x) > double.mant_dig * cast(double)LN2)
+    {
+        return copysign(1, x);
+    }
+    
+    const double y = expm1(2*x);
+    return y / (y + 2);
+}
 
 /// ditto
-float tanh(float x) @safe pure nothrow @nogc { return tanh(cast(real) x); }
+float tanh(float x) @safe pure nothrow @nogc
+{
+    if (fabs(x) > float.mant_dig * cast(float)LN2)
+    {
+        return copysign(1, x);
+    }
+    
+    const float y = expm1(2*x);
+    return y / (y + 2);
+}
```

Comment:
I think a private impl template would be much nicer than duplicating the body 3 times.